File: packages/shared/src/types/message.ts
import { z } from 'zod'

export const MessageType = z.enum(['text', 'structured', 'file', 'system'])
export type MessageType = z.infer<typeof MessageType>

export const MessageStatus = z.enum(['stored', 'delivered', 'read'])
export type MessageStatus = z.infer<typeof MessageStatus>

export const MessageContent = z.object({
  text: z.string().optional(),
  data: z.record(z.unknown()).optional(),
  // Attachment id returned by POST /v1/uploads. Lets the recipient call
  // GET /v1/attachments/:id to fetch the file. Keeping it as a top-level
  // field (instead of a convention inside `data`) makes it introspectable
  // by middleware, dashboards, and the sync drain path without having to
  // parse every agent's ad-hoc data shape. It's a transport-level access
  // control handle, not a description of the payload — which is why the
  // platform carries it but carries no file_url/file_name/mime_type.
  attachment_id: z.string().optional(),
}).refine(
  (c) =>
    c.text !== undefined ||
    c.data !== undefined ||
    c.attachment_id !== undefined,
  {
    message:
      'Message content must include at least one of: text, data, attachment_id',
  },
)
export type MessageContent = z.infer<typeof MessageContent>

export const Message = z.object({
  id: z.string(),
  conversation_id: z.string(),
  sender: z.string(),
  client_msg_id: z.string(),
  seq: z.number().int().nonnegative(),
  type: MessageType,
  content: MessageContent,
  metadata: z.record(z.unknown()).default({}),
  status: MessageStatus,
  created_at: z.string().datetime(),
  delivered_at: z.string().datetime().nullable(),
  read_at: z.string().datetime().nullable(),
})
export type Message = z.infer<typeof Message>

// Exactly one of `to` or `conversation_id` must be set. `to` is the
// direct-send path (handle resolution, cold-cap + block + inbox_mode
// gating, auto-create direct conversation). `conversation_id` is the
// group-send path (group membership already is the consent, so cold cap
// and block/inbox_mode checks are skipped — only per-second rate limit,
// payload size, and idempotency apply). Direct conversations are always
// addressed by handle via `to`; the group path is rejected with
// VALIDATION_ERROR if the id refers to a direct conversation.
export const SendMessageRequest = z.object({
  to: z.string().optional(),
  conversation_id: z.string().optional(),
  // Sender-provided idempotency key. Reusing the same value for this sender
  // returns the existing message instead of creating a duplicate. Generate a
  // UUID/ULID per logical send and retry with the same value on failure.
  client_msg_id: z.string().min(1).max(128),
  type: MessageType.default('text'),
  content: MessageContent,
  metadata: z.record(z.unknown()).optional(),
}).refine(
  (r) => (r.to !== undefined) !== (r.conversation_id !== undefined),
  { message: 'Exactly one of `to` or `conversation_id` must be provided' },
)
export type SendMessageRequest = z.infer<typeof SendMessageRequest>

File: packages/sdk-python/src/agentchat/types/message.py
"""Message wire types."""

from __future__ import annotations

from typing import Any, Dict, Literal, Optional

from pydantic import BaseModel, ConfigDict

MessageType = Literal["text", "structured", "file", "system"]
MessageStatus = Literal["stored", "delivered", "read"]


class _BaseModel(BaseModel):
    model_config = ConfigDict(extra="allow", populate_by_name=True)


class MessageContent(_BaseModel):
    """Payload body. At least one of ``text``, ``data``, or ``attachment_id`` must be set."""

    text: Optional[str] = None
    data: Optional[Dict[str, Any]] = None
    attachment_id: Optional[str] = None


class Message(_BaseModel):
    id: str
    conversation_id: str
    sender: str
    client_msg_id: str
    seq: int
    type: MessageType
    content: MessageContent
    metadata: Dict[str, Any] = {}
    status: MessageStatus
    created_at: str
    delivered_at: Optional[str] = None
    read_at: Optional[str] = None


class SendMessageRequest(_BaseModel):
    """Exactly one of ``to`` or ``conversation_id`` must be set.

    ``client_msg_id`` is the sender-supplied idempotency key. Reusing the
    same value returns the existing message instead of creating a
    duplicate. Generate a fresh UUID/ULID per logical send and reuse it
    on retry.
    """

    client_msg_id: str
    content: MessageContent
    to: Optional[str] = None
    conversation_id: Optional[str] = None
    type: Optional[MessageType] = None
    metadata: Optional[Dict[str, Any]] = None
